File: src/main/java/recipeapplication/service/UserService.java
package recipeapplication.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import recipeapplication.dto.UserDto;
import recipeapplication.exception.InvalidPasswordTokenException;
import recipeapplication.exception.UserNotFoundException;
import recipeapplication.model.PasswordResetToken;
import recipeapplication.model.User;
import recipeapplication.repository.PasswordTokenRepository;
import recipeapplication.repository.UserRepository;
import recipeapplication.security.RecipeUserDetails;
import recipeapplication.security.Role;

import java.util.Optional;
import java.util.UUID;

@Service
public class UserService
{
    private UserRepository userRepository;
    private PasswordTokenRepository passwordTokenRepository;
    private EmailService emailService;
    private PasswordEncoder passwordEncoder;
    private AuthService authService;

    @Autowired
    public UserService(
            UserRepository userRepository,
            PasswordTokenRepository passwordTokenRepository,
            EmailService emailService,
            PasswordEncoder passwordEncoder,
            AuthService authService)
    {
        this.userRepository = userRepository;
        this.passwordTokenRepository = passwordTokenRepository;
        this.emailService = emailService;
        this.passwordEncoder = passwordEncoder;
        this.authService = authService;
    }

    public void createPasswordResetToken(String serverName, UserDto userDto) throws UserNotFoundException
    {
        Optional<User> result;

        if (userDto.getUsername() != null)
        {
            result = userRepository.findByUsername(userDto.getUsername());
        }
        else if (userDto.getEmail() != null)
        {
            result = userRepository.findByEmail(userDto.getEmail());
        }
        else
        {
            throw new UserNotFoundException();
        }

        if (!result.isPresent())
        {
            throw new UserNotFoundException();
        }

        String token = UUID.randomUUID().toString();

        PasswordResetToken passwordResetToken = new PasswordResetToken(token, result.get());

        passwordTokenRepository.save(passwordResetToken);

        emailService.sendPasswordReset(result.get(), token, serverName);
    }

    public void processPasswordResetToken(String token) throws InvalidPasswordTokenException
    {
        Optional<PasswordResetToken> passwordResetToken = passwordTokenRepository.findByToken(token);

        if (!passwordResetToken.isPresent() || passwordResetToken.get().isExpired())
        {
            throw new InvalidPasswordTokenException();
        }

        User user = passwordResetToken.get().getUser();
        RecipeUserDetails recipeUserDetails = new RecipeUserDetails(user);
        recipeUserDetails.setChangePasswordAccess(true);

        authService.authenticateUser(recipeUserDetails, Role.CHANGE_PASSWORD);
    }

    public void changePassword(User user, String password)
    {
        user.setPassword(
                passwordEncoder.encode(password)
        );

        userRepository.save(user);
        passwordTokenRepository.deleteByUser(user);

        authService.disablePasswordReset();
    }

    public User getUser(String username) throws UserNotFoundException
    {
        Optional<User> user = userRepository.findByUsername(username);

        if (!user.isPresent())
        {
            throw new UserNotFoundException();
        }

        return user.get();
    }

    public void deleteAccount()
    {
        User user = authService.getLoggedInUser();

        userRepository.delete(user);
    }

    public void turnOffInstructions()
    {
        User user = authService.getLoggedInUser();

        if (user.isNewUser())
        {
            user.setNewUser(false);
            userRepository.save(user);
        }
    }
}

File: src/main/kotlin/recipeapplication/service/UserService.kt
package recipeapplication.service

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import recipeapplication.dto.PasswordResetDto
import recipeapplication.exception.InvalidPasswordTokenException
import recipeapplication.exception.UserNotFoundException
import recipeapplication.model.PasswordResetToken
import recipeapplication.model.User
import recipeapplication.repository.PasswordTokenRepository
import recipeapplication.repository.UserRepository
import recipeapplication.security.RecipeUserDetails
import recipeapplication.security.Role
import java.util.*
import kotlin.jvm.Throws

@Service
class UserService @Autowired constructor(
        private val userRepository: UserRepository,
        private val passwordTokenRepository: PasswordTokenRepository,
        private val emailService: EmailService,
        private val passwordEncoder: PasswordEncoder,
        private val authService: AuthService
) {
    @Throws(UserNotFoundException::class)
    fun createPasswordResetToken(serverName: String, passwordResetDto: PasswordResetDto) {
        val result = when {
            passwordResetDto.username != null -> userRepository.findByUsername(passwordResetDto.username!!)
            passwordResetDto.email != null -> userRepository.findByEmail(passwordResetDto.email!!)
            else -> throw UserNotFoundException()
        }

        if (!result.isPresent) {
            throw UserNotFoundException()
        }

        val token = UUID.randomUUID().toString()
        val passwordResetToken = PasswordResetToken(token, result.get())
        passwordTokenRepository.save(passwordResetToken)
        emailService.sendPasswordReset(result.get(), token, serverName)
    }

    @Throws(InvalidPasswordTokenException::class)
    fun processPasswordResetToken(token: String) {
        val passwordResetToken = passwordTokenRepository.findByToken(token)

        if (!passwordResetToken.isPresent || passwordResetToken.get().isExpired) {
            throw InvalidPasswordTokenException()
        }

        val user = passwordResetToken.get().user
        val recipeUserDetails = RecipeUserDetails(user)
        recipeUserDetails.changePasswordAccess = true
        authService.authenticateUser(recipeUserDetails, Role.CHANGE_PASSWORD)
    }

    fun changePassword(user: User, password: String) {
        user.password = passwordEncoder.encode(password)
        userRepository.save(user)
        passwordTokenRepository.deleteByUser(user)
        authService.disablePasswordReset()
    }

    @Throws(UserNotFoundException::class) // TODO can I get rid of these when the client is Kotlin?
    fun getUser(username: String): User {
        val user = userRepository.findByUsername(username)
        if (!user.isPresent) {
            throw UserNotFoundException()
        }
        return user.get()
    }

    fun deleteAccount() {
        userRepository.delete(
                authService.loggedInUser
        )
    }

    fun turnOffInstructions() {
        val user = authService.loggedInUser
        if (user.newUser) {
            user.newUser = false
            userRepository.save(user)
        }
    }

}